src/Snippets.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bolt;
+
+use Bolt\Snippet\QueueProcessor;
+use Bolt\Snippet\Target;
+use Bolt\Snippet\Zone;
+use Bolt\Widget\BaseWidget;
+use Bolt\Widget\BoltHeaderWidget;
+use Bolt\Widget\CanonicalLinkWidget;
+use Bolt\Widget\NewsWidget;
+use Bolt\Widget\WeatherWidget;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Tightenco\Collect\Support\Collection;
+use Twig\Environment;
+
+class Snippets
+{
+    /** @var Collection */
+    private $queue;
+
+    /** @var Request */
+    private $request;
+
+    /** @var QueueProcessor */
+    private $queueProcessor;
+
+    public function __construct(RequestStack $requestStack, QueueProcessor $queueProcessor)
+    {
+        $this->queue = new Collection([]);
+        $this->request = $requestStack->getCurrentRequest();
+        $this->queueProcessor = $queueProcessor;
+    }
+
+    /**
+     * @param BaseWidget|string|callable $callback
+     */
+    public function registerSnippet(
+        $callback,
+        string $target = Target::NOWHERE,
+        string $zone = Zone::FRONTEND,
+        string $name = 'nameless snippet',
+        int $priority = 100
+    ): void {
+        $this->queue->push([
+            'priority' => $priority,
+            'target' => $target,
+            'name' => $name,
+            'zone' => $zone,
+            'callback' => $callback,
+        ]);
+    }
+
+    public function registerWidget(BaseWidget $widget): void
+    {
+        $widget->setRequest($this->request);
```

Comment:
current request can get changed during page rendering execution so it can't be set here

Reply:
Since objects are references by default, if the Request changes, the widget using it will still "see" the latest, right?